service/src/main/java/uk/gov/hmcts/reform/fpl/service/HearingJudgeService.java

```diff
@@ -0,0 +1,211 @@
+package uk.gov.hmcts.reform.fpl.service;
+
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.lang3.tuple.Pair;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+import uk.gov.hmcts.reform.am.model.RoleAssignment;
+import uk.gov.hmcts.reform.fpl.model.HearingBooking;
+import uk.gov.hmcts.reform.fpl.model.common.Element;
+import uk.gov.hmcts.reform.rd.model.JudicialUserProfile;
+
+import java.time.Instant;
+import java.time.ZonedDateTime;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.UUID;
+
+import static org.apache.commons.lang3.ObjectUtils.isNotEmpty;
+import static org.springframework.util.ObjectUtils.isEmpty;
+import static uk.gov.hmcts.reform.fpl.config.TimeConfiguration.LONDON_TIMEZONE;
+import static uk.gov.hmcts.reform.fpl.enums.JudgeCaseRole.HEARING_JUDGE;
+import static uk.gov.hmcts.reform.fpl.enums.LegalAdviserRole.HEARING_LEGAL_ADVISER;
+
+@Slf4j
+@Service
+@RequiredArgsConstructor(onConstructor = @__(@Autowired))
+public class HearingJudgeService {
+
+    private final RoleAssignmentService roleAssignmentService;
+    private final JudicialService judicialService;
+
+    public static final List<String> HEARING_ROLES = List.of(
+        HEARING_JUDGE.getRoleName(),
+        HEARING_LEGAL_ADVISER.getRoleName());
+
+    /**
+     * Assign a hearing-judge on a case.
+     *
+     * @param caseId   the case id to add a case role on
+     * @param userId   the user to assign hearing-judge to
+     * @param starting the time to start the new role at, and the old roles to END at (- HEARING_EXPIRY_OFFSET_MINS)
+     */
+    private void assignHearingJudgeRole(Long caseId, String userId, ZonedDateTime starting, ZonedDateTime ending,
+                                        boolean isLegalAdviser) {
+        if (isLegalAdviser) {
+            roleAssignmentService.assignLegalAdvisersRole(caseId, List.of(userId), HEARING_LEGAL_ADVISER, starting,
+                ending);
+        } else {
+            roleAssignmentService.assignJudgesRole(caseId, List.of(userId), HEARING_JUDGE, starting, ending);
+        }
+    }
+
+    private void assignHearingJudge(Long caseId, HearingBooking hearing, Optional<HearingBooking> next) {
+        Optional<String> judgeId = getJudgeIdFromHearing(hearing);
+
+        // end this new role at the start of the next hearing (if present) MINUS 5 minutes to avoid overlapping roles
+        ZonedDateTime possibleEndDate = next.map(HearingBooking::getStartDate)
```

Comment:
probably just me not seeing it but are we missing the "MINUS 5 minutes" bit here?